fendermint/vm/topdown/src/voting.rs

```diff
@@ -0,0 +1,249 @@
+        &self,
+        block_height: BlockHeight,
+        block_hash: Option<V>,
+    ) -> StmResult<(), Error<K>> {
+        let mut chain = self.chain.read_clone()?;
+
+        // Check that we are extending the chain. We could also ignore existing heights.
+        if let Some((prev, _)) = chain.get_max() {
+            if block_height != prev + 1 {
+                return abort(Error::UnexpectedBlock(prev + 1, block_height));
+            }
+        }
+
+        chain.insert(block_height, block_hash);
+
+        self.chain.write(chain)?;
+
+        Ok(())
+    }
+
+    /// Add a vote we received.
+    ///
+    /// Returns `true` if this vote was added, `false` if it was ignored as a
+    /// duplicate or a height we already finalized, and an error if it's an
+    /// equivocation or from a validator we don't know.
+    pub fn add_vote(
+        &self,
+        validator_key: K,
+        block_height: BlockHeight,
+        block_hash: V,
+    ) -> StmResult<bool, Error<K, V>> {
+        let min_height = self.last_finalized_height()?;
+
+        if block_height < min_height {
+            return Ok(false);
+        }
+
+        if !self.known_validator(&validator_key)? {
+            return abort(Error::UnknownValidator(validator_key));
+        }
+
+        let mut votes = self.votes.read_clone()?;
+        let votes_at_height = votes.entry(block_height).or_default();
+
+        for (bh, vs) in votes_at_height.iter() {
+            if *bh != block_hash && vs.contains(&validator_key) {
+                return abort(Error::Equivocation(
+                    validator_key,
+                    block_height,
+                    block_hash,
+                    bh.clone(),
+                ));
+            }
+        }
+
+        let votes_for_block = votes_at_height.entry(block_hash).or_default();
+
+        if votes_for_block.insert(validator_key).is_some() {
+            return Ok(false);
+        }
+
+        if *self.pause_votes.read()? {
+            retry()?;
+        }
+
+        self.votes.write(votes)?;
+
+        Ok(true)
+    }
+
+    /// Pause adding more votes until we are finished calling `find_quorum` which
+    /// automatically re-enables them.
+    pub fn pause_votes_until_find_quorum(&self) -> Stm<()> {
+        self.pause_votes.write(true)
+    }
+
+    /// Find a block on the (from our perspective) finalized chain that gathered enough votes from validators.
+    pub fn find_quorum(&self) -> Stm<Option<(BlockHeight, V)>> {
+        self.pause_votes.write(false)?;
+
+        let quorum_threshold = self.quorum_threshold()?;
+        let chain = self.chain.read()?;
+
+        let Some((finalized_height, _)) = chain.get_min() else {
+            return Ok(None);
+        };
+
+        let votes = self.votes.read()?;
+        let power_table = self.power_table.read()?;
+
+        let mut weight = 0;
+        let mut voters = im::HashSet::new();
+
+        for (block_height, block_hash) in chain.iter().rev() {
```

Comment:
so you basically just look from oldest block to newest and add up the weights from all the voters until you exceed the power threshold, and at that height, we have found last common block that will pass validator proposal?

Reply:
Yes, except it's looking from the newest block back to the oldest, so that it can find quorum for the highest height possible.